File: iOS/StreamVision/Sources/Services/XtreamCodesService.swift
import Foundation

/// Service complet pour l'API Xtream Codes
/// Supporte Live, VOD, Series, Catch-up
actor XtreamCodesService {

    private let session: URLSession
    private var currentAccount: XtreamAccountInfo?

    init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        self.session = URLSession(configuration: config)
    }

    // MARK: - Authentication

    func authenticate(serverUrl: String, username: String, password: String) async throws -> XtreamAccountInfo? {
        let normalizedUrl = normalizeServerUrl(serverUrl)
        guard let url = URL(string: "\(normalizedUrl)/player_api.php?username=\(username)&password=\(password)") else {
            throw XtreamError.invalidURL
        }

        let (data, _) = try await session.data(from: url)

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let userInfo = json["user_info"] as? [String: Any] else {
            return nil
        }

        let account = XtreamAccountInfo(
            username: userInfo["username"] as? String ?? "",
            status: userInfo["status"] as? String ?? "",
            expDate: userInfo["exp_date"] as? String ?? "",
            maxConnections: userInfo["max_connections"] as? String ?? "",
            activeConnections: userInfo["active_cons"] as? String ?? "",
            serverUrl: normalizedUrl
        )
        currentAccount = account
        return account
    }

    // MARK: - Categories

    func getLiveCategories(serverUrl: String, username: String, password: String) async throws -> [XtreamCategory] {
        return try await getCategories(serverUrl: serverUrl, username: username, password: password, action: "get_live_categories")
    }

    func getVodCategories(serverUrl: String, username: String, password: String) async throws -> [XtreamCategory] {
        return try await getCategories(serverUrl: serverUrl, username: username, password: password, action: "get_vod_categories")
    }

    func getSeriesCategories(serverUrl: String, username: String, password: String) async throws -> [XtreamCategory] {
        return try await getCategories(serverUrl: serverUrl, username: username, password: password, action: "get_series_categories")
    }

    private func getCategories(serverUrl: String, username: String, password: String, action: String) async throws -> [XtreamCategory] {
        let normalizedUrl = normalizeServerUrl(serverUrl)
        guard let url = URL(string: "\(normalizedUrl)/player_api.php?username=\(username)&password=\(password)&action=\(action)") else {
            throw XtreamError.invalidURL
        }

        let (data, _) = try await session.data(from: url)

        guard let jsonArray = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return []
        }

        return jsonArray.compactMap { item in
            guard let categoryId = (item["category_id"] as? String) ?? (item["category_id"] as? Int).map(String.init),
                  let categoryName = item["category_name"] as? String else {
                return nil
            }
            return XtreamCategory(
                categoryId: categoryId,
                categoryName: categoryName,
                parentId: item["parent_id"] as? String ?? ""
            )
        }
    }

    // MARK: - Live Streams

    func getLiveStreams(serverUrl: String, username: String, password: String, sourceId: UUID) async throws -> [MediaItem] {
        let normalizedUrl = normalizeServerUrl(serverUrl)
        guard let url = URL(string: "\(normalizedUrl)/player_api.php?username=\(username)&password=\(password)&action=get_live_streams") else {
            throw XtreamError.invalidURL
        }

        let (data, _) = try await session.data(from: url)

        guard let jsonArray = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return []
        }

        let categories = try await getLiveCategories(serverUrl: serverUrl, username: username, password: password)
        let categoryDict = Dictionary(uniqueKeysWithValues: categories.map { ($0.categoryId, $0.categoryName) })

        var items: [MediaItem] = []
        var order = 0

        for item in jsonArray {
            guard let streamId = (item["stream_id"] as? Int) ?? Int(item["stream_id"] as? String ?? "") else {
                continue
            }

            let categoryId = (item["category_id"] as? String) ?? (item["category_id"] as? Int).map(String.init) ?? ""
            let archiveDuration = (item["tv_archive_duration"] as? Int) ?? Int(item["tv_archive_duration"] as? String ?? "") ?? 0

            let mediaItem = MediaItem(
                sourceId: sourceId,
                name: item["name"] as? String ?? "",
                mediaType: .live,
                logoUrl: item["stream_icon"] as? String,
                streamUrl: "\(normalizedUrl)/live/\(username)/\(password)/\(streamId).m3u8",
                groupTitle: categoryDict[categoryId] ?? "Uncategorized",
                categoryId: categoryId,
                epgId: item["epg_channel_id"] as? String,
                catchupDays: archiveDuration,
                order: order
            )

            items.append(mediaItem)
            order += 1
        }

        return items
    }

    // Compatibilite avec l'ancien code
    func getLiveStreamsAsChannels(serverUrl: String, username: String, password: String, sourceId: UUID) async throws -> [Channel] {
        let mediaItems = try await getLiveStreams(serverUrl: serverUrl, username: username, password: password, sourceId: sourceId)
        return mediaItems.map { m in
            Channel(
                id: m.id,
                sourceId: m.sourceId,
                name: m.name,
                logoUrl: m.logoUrl,
                streamUrl: m.streamUrl,
                groupTitle: m.groupTitle,
                epgId: m.epgId,
                catchupDays: m.catchupDays,
                order: m.order
            )
        }
    }

    // MARK: - VOD (Movies)

    func getVodStreams(serverUrl: String, username: String, password: String, sourceId: UUID, categoryId: String? = nil) async throws -> [MediaItem] {
        let normalizedUrl = normalizeServerUrl(serverUrl)
        var urlString = "\(normalizedUrl)/player_api.php?username=\(username)&password=\(password)&action=get_vod_streams"
        if let categoryId = categoryId, !categoryId.isEmpty {
            urlString += "&category_id=\(categoryId)"
        }

        guard let url = URL(string: urlString) else {
            throw XtreamError.invalidURL
        }

        let (data, _) = try await session.data(from: url)

        guard let jsonArray = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return []
        }

        let categories = try await getVodCategories(serverUrl: serverUrl, username: username, password: password)
        let categoryDict = Dictionary(uniqueKeysWithValues: categories.map { ($0.categoryId, $0.categoryName) })

        var items: [MediaItem] = []
        var order = 0

        for item in jsonArray {
            guard let streamId = (item["stream_id"] as? Int) ?? Int(item["stream_id"] as? String ?? "") else {
                continue
            }

            let extension_ = item["container_extension"] as? String ?? "mp4"
            let catId = (item["category_id"] as? String) ?? (item["category_id"] as? Int).map(String.init) ?? ""

            var mediaItem = MediaItem(
                sourceId: sourceId,
                name: item["name"] as? String ?? "",
                mediaType: .movie,
                posterUrl: item["stream_icon"] as? String,
                streamUrl: "\(normalizedUrl)/movie/\(username)/\(password)/\(streamId).\(extension_)",
                groupTitle: categoryDict[catId] ?? "Films",
                categoryId: catId,
                containerExtension: extension_,
                order: order,
                rating: item["rating"] as? Double ?? Double(item["rating"] as? String ?? "") ?? 0.0,
                tmdbId: Int(item["tmdb_id"] as? String ?? "")
            )

            if let added = item["added"] as? String, let timestamp = TimeInterval(added) {
                mediaItem.releaseDate = Date(timeIntervalSince1970: timestamp)
            }

            items.append(mediaItem)
            order += 1
        }

        return items
    }

    func getVodInfo(serverUrl: String, username: String, password: String, vodId: String) async throws -> VodInfo? {
        let normalizedUrl = normalizeServerUrl(serverUrl)
        guard let url = URL(string: "\(normalizedUrl)/player_api.php?username=\(username)&password=\(password)&action=get_vod_info&vod_id=\(vodId)") else {
            throw XtreamError.invalidURL
        }

        let (data, _) = try await session.data(from: url)

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }

        let info = json["info"] as? [String: Any]
        let movieData = json["movie_data"] as? [String: Any]

        return VodInfo(
            streamId: vodId,
            name: info?["name"] as? String ?? movieData?["name"] as? String ?? "",
            overview: info?["plot"] as? String ?? info?["description"] as? String ?? "",
            posterUrl: info?["movie_image"] as? String ?? info?["cover_big"] as? String,
            backdropUrl: (info?["backdrop_path"] as? [String])?.first,
            rating: info?["rating"] as? Double ?? Double(info?["rating"] as? String ?? "") ?? 0.0,
            duration: info?["duration"] as? String ?? "",
            releaseDate: info?["releasedate"] as? String ?? info?["release_date"] as? String ?? "",
            genre: info?["genre"] as? String ?? "",
            director: info?["director"] as? String ?? "",
            cast: info?["cast"] as? String ?? info?["actors"] as? String ?? "",
            tmdbId: Int(info?["tmdb_id"] as? String ?? ""),
            trailerUrl: info?["youtube_trailer"] as? String,
            containerExtension: movieData?["container_extension"] as? String ?? "mp4"
        )
    }

    // MARK: - Series

    func getSeries(serverUrl: String, username: String, password: String, sourceId: UUID, categoryId: String? = nil) async throws -> [MediaItem] {
        let normalizedUrl = normalizeServerUrl(serverUrl)
        var urlString = "\(normalizedUrl)/player_api.php?username=\(username)&password=\(password)&action=get_series"
        if let categoryId = categoryId, !categoryId.isEmpty {
            urlString += "&category_id=\(categoryId)"
        }

        guard let url = URL(string: urlString) else {
            throw XtreamError.invalidURL
        }

        let (data, _) = try await session.data(from: url)

        guard let jsonArray = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return []
        }

        let categories = try await getSeriesCategories(serverUrl: serverUrl, username: username, password: password)
        let categoryDict = Dictionary(uniqueKeysWithValues: categories.map { ($0.categoryId, $0.categoryName) })

        var items: [MediaItem] = []
        var order = 0

        for item in jsonArray {
            let seriesId = item["series_id"] as? String ?? "\(item["series_id"] as? Int ?? 0)"
            let catId = (item["category_id"] as? String) ?? (item["category_id"] as? Int).map(String.init) ?? ""

            let mediaItem = MediaItem(
                sourceId: sourceId,
                name: item["name"] as? String ?? "",
                mediaType: .series,
                posterUrl: item["cover"] as? String,
                backdropUrl: (item["backdrop_path"] as? [String])?.first,
                groupTitle: categoryDict[catId] ?? "Series",
                categoryId: catId,
                seriesId: seriesId,
                order: order,
                rating: item["rating"] as? Double ?? Double(item["rating"] as? String ?? "") ?? 0.0,
                tmdbId: Int(item["tmdb_id"] as? String ?? ""),
                overview: item["plot"] as? String,
                genres: item["genre"] as? String,
                cast: item["cast"] as? String
            )

            items.append(mediaItem)
            order += 1
        }

        return items
    }

    func getSeriesInfo(serverUrl: String, username: String, password: String, seriesId: String) async throws -> SeriesFullInfo? {
        let normalizedUrl = normalizeServerUrl(serverUrl)
        guard let url = URL(string: "\(normalizedUrl)/player_api.php?username=\(username)&password=\(password)&action=get_series_info&series_id=\(seriesId)") else {
            throw XtreamError.invalidURL
        }

        let (data, _) = try await session.data(from: url)

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }

        let info = json["info"] as? [String: Any]
        let episodes = json["episodes"] as? [String: [[String: Any]]]

        var seasons: [SeriesSeasonInfo] = []

        if let episodes = episodes {
            for (seasonKey, episodesArray) in episodes {
                let seasonNum = Int(seasonKey) ?? 0

                let episodeList: [EpisodeInfo] = episodesArray.compactMap { ep in
                    guard let episodeId = ep["id"] as? String ?? (ep["id"] as? Int).map(String.init) else {
                        return nil
                    }

                    let extension_ = ep["container_extension"] as? String ?? "mp4"

                    return EpisodeInfo(
                        id: episodeId,
                        episodeNumber: ep["episode_num"] as? Int ?? Int(ep["episode_num"] as? String ?? "") ?? 0,
                        title: ep["title"] as? String ?? "Episode",
                        overview: ep["plot"] as? String ?? "",
                        duration: ep["duration"] as? String ?? "",
                        posterUrl: (ep["info"] as? [String: Any])?["movie_image"] as? String,
                        streamUrl: "\(normalizedUrl)/series/\(username)/\(password)/\(episodeId).\(extension_)",
                        containerExtension: extension_,
                        rating: ep["rating"] as? Double ?? 0.0
                    )
                }

                seasons.append(SeriesSeasonInfo(
                    seasonNumber: seasonNum,
                    episodes: episodeList
                ))
            }
        }

        return SeriesFullInfo(
            seriesId: seriesId,
            name: info?["name"] as? String ?? "",
            overview: info?["plot"] as? String ?? "",
            posterUrl: info?["cover"] as? String,
            backdropUrl: (info?["backdrop_path"] as? [String])?.first,
            rating: info?["rating"] as? Double ?? 0.0,
            genre: info?["genre"] as? String ?? "",
            director: info?["director"] as? String ?? "",
            cast: info?["cast"] as? String ?? "",
            releaseDate: info?["releaseDate"] as? String ?? "",
            tmdbId: Int(info?["tmdb_id"] as? String ?? ""),
            seasons: seasons.sorted { $0.seasonNumber < $1.seasonNumber }
        )
    }

    // MARK: - Catch-up / Timeshift

    func getCatchupURL(serverUrl: String, username: String, password: String, streamId: String, startTime: Date, durationMinutes: Int) -> String {
        let normalizedUrl = normalizeServerUrl(serverUrl)
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd:HH-mm"
        formatter.timeZone = TimeZone(identifier: "UTC")
        let startUtc = formatter.string(from: startTime)

        return "\(normalizedUrl)/timeshift/\(username)/\(password)/\(durationMinutes)/\(startUtc)/\(streamId).m3u8"
    }

    func getCatchupURLSimple(serverUrl: String, username: String, password: String, streamId: String, startTime: Date) -> String {
        let normalizedUrl = normalizeServerUrl(serverUrl)
        let start = Int(startTime.timeIntervalSince1970)

        return "\(normalizedUrl)/streaming/timeshift.php?username=\(username)&password=\(password)&stream=\(streamId)&start=\(start)"
    }

    // MARK: - Short EPG

    func getShortEpg(serverUrl: String, username: String, password: String, streamId: String, limit: Int = 10) async throws -> [XtreamEpgEntry] {
        let normalizedUrl = normalizeServerUrl(serverUrl)
        guard let url = URL(string: "\(normalizedUrl)/player_api.php?username=\(username)&password=\(password)&action=get_short_epg&stream_id=\(streamId)&limit=\(limit)") else {
            throw XtreamError.invalidURL
        }

        let (data, _) = try await session.data(from: url)

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let epgListings = json["epg_listings"] as? [[String: Any]] else {
            return []
        }

        return epgListings.compactMap { item in
            XtreamEpgEntry(
                id: item["id"] as? String ?? "",
                title: decodeBase64(item["title"] as? String ?? ""),
                description: decodeBase64(item["description"] as? String ?? ""),
                start: parseEpgDateTime(item["start"] as? String),
                end: parseEpgDateTime(item["end"] as? String),
                startTimestamp: item["start_timestamp"] as? String ?? "",
                stopTimestamp: item["stop_timestamp"] as? String ?? ""
            )
        }
    }

    // MARK: - Helpers

    private func normalizeServerUrl(_ url: String) -> String {
        var normalized = url.trimmingCharacters(in: .whitespacesAndNewlines)
        if normalized.hasSuffix("/") {
            normalized = String(normalized.dropLast())
        }
        if !normalized.hasPrefix("http://") && !normalized.hasPrefix("https://") {
            normalized = "http://\(normalized)"
        }
        return normalized
    }

    private func decodeBase64(_ base64: String) -> String {
        guard !base64.isEmpty,
              let data = Data(base64Encoded: base64),
              let decoded = String(data: data, encoding: .utf8) else {
            return base64
        }
        return decoded
    }

    private func parseEpgDateTime(_ dateStr: String?) -> Date? {
        guard let dateStr = dateStr, !dateStr.isEmpty else { return nil }
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
        return formatter.date(from: dateStr)
    }
}

// MARK: - Errors

enum XtreamError: Error {
    case invalidURL
    case authenticationFailed
    case networkError(Error)
}

File: Android/app/src/main/java/com/streamvision/app/data/repository/XtreamCodesService.kt
package com.streamvision.app.data.repository

import android.util.Base64
import com.streamvision.app.data.models.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Service complet pour l'API Xtream Codes
 * Supporte Live, VOD, Series, Catch-up
 */
@Singleton
class XtreamCodesService @Inject constructor() {

    private var currentAccount: XtreamAccountInfo? = null

    // region Authentication

    suspend fun authenticate(
        serverUrl: String,
        username: String,
        password: String
    ): XtreamAccountInfo? = withContext(Dispatchers.IO) {
        try {
            val normalizedUrl = normalizeServerUrl(serverUrl)
            val url = "$normalizedUrl/player_api.php?username=$username&password=$password"
            val response = URL(url).readText()
            val json = JSONObject(response)

            val userInfo = json.optJSONObject("user_info") ?: return@withContext null
            val serverInfo = json.optJSONObject("server_info")

            currentAccount = XtreamAccountInfo(
                username = userInfo.optString("username", ""),
                status = userInfo.optString("status", ""),
                expDate = userInfo.optString("exp_date", ""),
                maxConnections = userInfo.optString("max_connections", ""),
                activeConnections = userInfo.optString("active_cons", ""),
                serverUrl = normalizedUrl
            )
            currentAccount
        } catch (e: Exception) {
            null
        }
    }

    // endregion

    // region Categories

    suspend fun getLiveCategories(
        serverUrl: String,
        username: String,
        password: String
    ): List<XtreamCategory> = getCategories(serverUrl, username, password, "get_live_categories")

    suspend fun getVodCategories(
        serverUrl: String,
        username: String,
        password: String
    ): List<XtreamCategory> = getCategories(serverUrl, username, password, "get_vod_categories")

    suspend fun getSeriesCategories(
        serverUrl: String,
        username: String,
        password: String
    ): List<XtreamCategory> = getCategories(serverUrl, username, password, "get_series_categories")

    private suspend fun getCategories(
        serverUrl: String,
        username: String,
        password: String,
        action: String
    ): List<XtreamCategory> = withContext(Dispatchers.IO) {
        try {
            val normalizedUrl = normalizeServerUrl(serverUrl)
            val url = "$normalizedUrl/player_api.php?username=$username&password=$password&action=$action"
            val response = URL(url).readText()
            val jsonArray = JSONArray(response)

            (0 until jsonArray.length()).mapNotNull { i ->
                val item = jsonArray.getJSONObject(i)
                XtreamCategory(
                    categoryId = item.optString("category_id", ""),
                    categoryName = item.optString("category_name", ""),
                    parentId = item.optString("parent_id", "")
                )
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    // endregion

    // region Live Streams

    suspend fun getLiveStreams(
        serverUrl: String,
        username: String,
        password: String,
        sourceId: String
    ): List<MediaItem> = withContext(Dispatchers.IO) {
        try {
            val normalizedUrl = normalizeServerUrl(serverUrl)
            val url = "$normalizedUrl/player_api.php?username=$username&password=$password&action=get_live_streams"
            val response = URL(url).readText()
            val jsonArray = JSONArray(response)

            val categories = getLiveCategories(serverUrl, username, password)
            val categoryMap = categories.associateBy { it.categoryId }

            (0 until jsonArray.length()).mapNotNull { i ->
                val item = jsonArray.getJSONObject(i)
                val streamId = item.optString("stream_id", "")
                val categoryId = item.optString("category_id", "")
                val catchupDays = item.optInt("tv_archive_duration", 0)

                MediaItem(
                    sourceId = sourceId,
                    name = item.optString("name", ""),
                    logoUrl = item.optString("stream_icon", null),
                    streamUrl = "$normalizedUrl/live/$username/$password/$streamId.m3u8",
                    groupTitle = categoryMap[categoryId]?.categoryName ?: "Uncategorized",
                    categoryId = categoryId,
                    epgId = item.optString("epg_channel_id", null),
                    catchupDays = catchupDays,
                    order = i,
                    mediaType = ContentType.LIVE
                )
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    // Compatibilite avec l'ancien code
    suspend fun getLiveStreamsAsChannels(
        serverUrl: String,
        username: String,
        password: String,
        sourceId: String
    ): List<Channel> {
        val mediaItems = getLiveStreams(serverUrl, username, password, sourceId)
        return mediaItems.map { m ->
            Channel(
                id = m.id,
                sourceId = m.sourceId,
                name = m.name,
                logoUrl = m.logoUrl,
                streamUrl = m.streamUrl,
                groupTitle = m.groupTitle,
                epgId = m.epgId,
                catchupDays = m.catchupDays,
                order = m.order
            )
        }
    }

    // endregion

    // region VOD (Movies)

    suspend fun getVodStreams(
        serverUrl: String,
        username: String,
        password: String,
        sourceId: String,
        categoryId: String? = null
    ): List<MediaItem> = withContext(Dispatchers.IO) {
        try {
            val normalizedUrl = normalizeServerUrl(serverUrl)
            var url = "$normalizedUrl/player_api.php?username=$username&password=$password&action=get_vod_streams"
            if (!categoryId.isNullOrEmpty()) {
                url += "&category_id=$categoryId"
            }

            val response = URL(url).readText()
            val jsonArray = JSONArray(response)

            val categories = getVodCategories(serverUrl, username, password)
            val categoryMap = categories.associateBy { it.categoryId }

            (0 until jsonArray.length()).mapNotNull { i ->
                val item = jsonArray.getJSONObject(i)
                val streamId = item.optString("stream_id", "")
                val extension = item.optString("container_extension", "mp4")
                val catId = item.optString("category_id", "")

                MediaItem(
                    sourceId = sourceId,
                    name = item.optString("name", ""),
                    posterUrl = item.optString("stream_icon", null),
                    streamUrl = "$normalizedUrl/movie/$username/$password/$streamId.$extension",
                    groupTitle = categoryMap[catId]?.categoryName ?: "Films",
                    categoryId = catId,
                    order = i,
                    mediaType = ContentType.MOVIE,
                    containerExtension = extension,
                    rating = item.optDouble("rating", 0.0),
                    tmdbId = item.optString("tmdb_id", null)?.toIntOrNull(),
                    releaseDate = item.optString("added", null)?.toLongOrNull()?.let { it * 1000 }
                )
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun getVodInfo(
        serverUrl: String,
        username: String,
        password: String,
        vodId: String
    ): VodInfo? = withContext(Dispatchers.IO) {
        try {
            val normalizedUrl = normalizeServerUrl(serverUrl)
            val url = "$normalizedUrl/player_api.php?username=$username&password=$password&action=get_vod_info&vod_id=$vodId"
            val response = URL(url).readText()
            val json = JSONObject(response)

            val info = json.optJSONObject("info")
            val movieData = json.optJSONObject("movie_data")

            VodInfo(
                streamId = vodId,
                name = info?.optString("name") ?: movieData?.optString("name") ?: "",
                overview = info?.optString("plot") ?: info?.optString("description") ?: "",
                posterUrl = info?.optString("movie_image") ?: info?.optString("cover_big"),
                backdropUrl = info?.optJSONArray("backdrop_path")?.optString(0),
                rating = info?.optDouble("rating", 0.0) ?: 0.0,
                duration = info?.optString("duration") ?: "",
                releaseDate = info?.optString("releasedate") ?: info?.optString("release_date") ?: "",
                genre = info?.optString("genre") ?: "",
                director = info?.optString("director") ?: "",
                cast = info?.optString("cast") ?: info?.optString("actors") ?: "",
                tmdbId = info?.optString("tmdb_id")?.toIntOrNull(),
                trailerUrl = info?.optString("youtube_trailer"),
                containerExtension = movieData?.optString("container_extension") ?: "mp4"
            )
        } catch (e: Exception) {
            null
        }
    }

    // endregion

    // region Series

    suspend fun getSeries(
        serverUrl: String,
        username: String,
        password: String,
        sourceId: String,
        categoryId: String? = null
    ): List<MediaItem> = withContext(Dispatchers.IO) {
        try {
            val normalizedUrl = normalizeServerUrl(serverUrl)
            var url = "$normalizedUrl/player_api.php?username=$username&password=$password&action=get_series"
            if (!categoryId.isNullOrEmpty()) {
                url += "&category_id=$categoryId"
            }

            val response = URL(url).readText()
            val jsonArray = JSONArray(response)

            val categories = getSeriesCategories(serverUrl, username, password)
            val categoryMap = categories.associateBy { it.categoryId }

            (0 until jsonArray.length()).mapNotNull { i ->
                val item = jsonArray.getJSONObject(i)
                val seriesId = item.optString("series_id", "")
                val catId = item.optString("category_id", "")

                MediaItem(
                    sourceId = sourceId,
                    seriesId = seriesId,
                    name = item.optString("name", ""),
                    posterUrl = item.optString("cover", null),
                    backdropUrl = item.optJSONArray("backdrop_path")?.optString(0),
                    streamUrl = "", // Les series n'ont pas d'URL directe
                    groupTitle = categoryMap[catId]?.categoryName ?: "Series",
                    categoryId = catId,
                    order = i,
                    mediaType = ContentType.SERIES,
                    rating = item.optDouble("rating", 0.0),
                    tmdbId = item.optString("tmdb_id", null)?.toIntOrNull(),
                    overview = item.optString("plot", ""),
                    genres = item.optString("genre", ""),
                    cast = item.optString("cast", "")
                )
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun getSeriesInfo(
        serverUrl: String,
        username: String,
        password: String,
        seriesId: String
    ): SeriesFullInfo? = withContext(Dispatchers.IO) {
        try {
            val normalizedUrl = normalizeServerUrl(serverUrl)
            val url = "$normalizedUrl/player_api.php?username=$username&password=$password&action=get_series_info&series_id=$seriesId"
            val response = URL(url).readText()
            val json = JSONObject(response)

            val info = json.optJSONObject("info")
            val episodes = json.optJSONObject("episodes")

            val seasons = mutableListOf<SeriesSeasonInfo>()

            if (episodes != null) {
                val seasonKeys = episodes.keys()
                while (seasonKeys.hasNext()) {
                    val seasonKey = seasonKeys.next()
                    val seasonNum = seasonKey.toIntOrNull() ?: 0
                    val episodesArray = episodes.optJSONArray(seasonKey)

                    val episodeList = mutableListOf<EpisodeInfo>()
                    if (episodesArray != null) {
                        for (j in 0 until episodesArray.length()) {
                            val ep = episodesArray.getJSONObject(j)
                            val episodeId = ep.optString("id", "")
                            val extension = ep.optString("container_extension", "mp4")

                            episodeList.add(EpisodeInfo(
                                id = episodeId,
                                episodeNumber = ep.optInt("episode_num", 0),
                                title = ep.optString("title", "Episode"),
                                overview = ep.optString("plot", ""),
                                duration = ep.optString("duration", ""),
                                posterUrl = ep.optJSONObject("info")?.optString("movie_image"),
                                streamUrl = "$normalizedUrl/series/$username/$password/$episodeId.$extension",
                                containerExtension = extension,
                                rating = ep.optDouble("rating", 0.0)
                            ))
                        }
                    }

                    seasons.add(SeriesSeasonInfo(
                        seasonNumber = seasonNum,
                        episodes = episodeList
                    ))
                }
            }

            SeriesFullInfo(
                seriesId = seriesId,
                name = info?.optString("name") ?: "",
                overview = info?.optString("plot") ?: "",
                posterUrl = info?.optString("cover"),
                backdropUrl = info?.optJSONArray("backdrop_path")?.optString(0),
                rating = info?.optDouble("rating", 0.0) ?: 0.0,
                genre = info?.optString("genre") ?: "",
                director = info?.optString("director") ?: "",
                cast = info?.optString("cast") ?: "",
                releaseDate = info?.optString("releaseDate") ?: "",
                tmdbId = info?.optString("tmdb_id")?.toIntOrNull(),
                seasons = seasons.sortedBy { it.seasonNumber }
            )
        } catch (e: Exception) {
            null
        }
    }

    // endregion

    // region Catch-up / Timeshift

    fun getCatchupUrl(
        serverUrl: String,
        username: String,
        password: String,
        streamId: String,
        startTime: Date,
        durationMinutes: Int
    ): String {
        val normalizedUrl = normalizeServerUrl(serverUrl)
        val format = SimpleDateFormat("yyyy-MM-dd:HH-mm", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }
        val startUtc = format.format(startTime)
        return "$normalizedUrl/timeshift/$username/$password/$durationMinutes/$startUtc/$streamId.m3u8"
    }

    fun getCatchupUrlSimple(
        serverUrl: String,
        username: String,
        password: String,
        streamId: String,
        startTime: Date
    ): String {
        val normalizedUrl = normalizeServerUrl(serverUrl)
        val start = startTime.time / 1000
        return "$normalizedUrl/streaming/timeshift.php?username=$username&password=$password&stream=$streamId&start=$start"
    }

    // endregion

    // region Short EPG

    suspend fun getShortEpg(
        serverUrl: String,
        username: String,
        password: String,
        streamId: String,
        limit: Int = 10
    ): List<XtreamEpgEntry> = withContext(Dispatchers.IO) {
        try {
            val normalizedUrl = normalizeServerUrl(serverUrl)
            val url = "$normalizedUrl/player_api.php?username=$username&password=$password&action=get_short_epg&stream_id=$streamId&limit=$limit"
            val response = URL(url).readText()
            val json = JSONObject(response)

            val epgListings = json.optJSONArray("epg_listings") ?: return@withContext emptyList()

            (0 until epgListings.length()).mapNotNull { i ->
                val item = epgListings.getJSONObject(i)
                XtreamEpgEntry(
                    id = item.optString("id", ""),
                    title = decodeBase64(item.optString("title", "")),
                    description = decodeBase64(item.optString("description", "")),
                    start = parseEpgDateTime(item.optString("start")),
                    end = parseEpgDateTime(item.optString("end")),
                    startTimestamp = item.optString("start_timestamp", ""),
                    stopTimestamp = item.optString("stop_timestamp", "")
                )
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    // endregion

    // region Helpers

    private fun normalizeServerUrl(url: String): String {
        var normalized = url.trim().trimEnd('/')
        if (!normalized.startsWith("http://") && !normalized.startsWith("https://")) {
            normalized = "http://$normalized"
        }
        return normalized
    }

    private fun decodeBase64(base64: String): String {
        if (base64.isEmpty()) return ""
        return try {
            String(Base64.decode(base64, Base64.DEFAULT), Charsets.UTF_8)
        } catch (e: Exception) {
            base64 // Return as-is if not base64
        }
    }

    private fun parseEpgDateTime(dateStr: String?): Long? {
        if (dateStr.isNullOrEmpty()) return null
        return try {
            SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US).parse(dateStr)?.time
        } catch (e: Exception) {
            null
        }
    }

    // endregion
}
